provider/ultradns_test.go

```diff
@@ -0,0 +1,658 @@
+                              "owner":"teamrest",
+                              "resourceRecordCount":7,
+                              "lastModifiedDateTime":""
+                           }
+                        }`
+	if err := json.Unmarshal([]byte(zoneJson), &zone); err != nil {
+		log.Fatal(err)
+	}
+
+	zones = append(zones, zone)
+	return zones, udnssdk.ResultInfo{}, nil, nil
+}
+
+type mockUltraDNSRecord struct {
+	client *udnssdk.Client
+}
+
+func (m *mockUltraDNSRecord) Create(k udnssdk.RRSetKey, rrset udnssdk.RRSet) (*http.Response, error) {
+	return nil, nil
+}
+
+func (m *mockUltraDNSRecord) Select(k udnssdk.RRSetKey) ([]udnssdk.RRSet, error) {
+	return []udnssdk.RRSet{{
+		OwnerName: "test-ultradns-provider.com.",
+		RRType:    endpoint.RecordTypeA,
+		RData:     []string{"1.1.1.1"},
+		TTL:       86400,
+	}}, nil
+
+}
+
+func (m *mockUltraDNSRecord) SelectWithOffset(k udnssdk.RRSetKey, offset int) ([]udnssdk.RRSet, udnssdk.ResultInfo, *http.Response, error) {
+	return nil, udnssdk.ResultInfo{}, nil, nil
+}
+
+func (m *mockUltraDNSRecord) Update(udnssdk.RRSetKey, udnssdk.RRSet) (*http.Response, error) {
+	return nil, nil
+}
+
+func (m *mockUltraDNSRecord) Delete(k udnssdk.RRSetKey) (*http.Response, error) {
+	return nil, nil
+}
+
+func (m *mockUltraDNSRecord) SelectWithOffsetWithLimit(k udnssdk.RRSetKey, offset int, limit int) (rrsets []udnssdk.RRSet, ResultInfo udnssdk.ResultInfo, resp *http.Response, err error) {
+	return []udnssdk.RRSet{{
+		OwnerName: "test-ultradns-provider.com.",
+		RRType:    endpoint.RecordTypeA,
+		RData:     []string{"1.1.1.1"},
+		TTL:       86400,
+	}}, udnssdk.ResultInfo{}, nil, nil
+}
+
+// NewUltraDNSProvider Test scenario
+func TestNewUltraDNSProvider(t *testing.T) {
+	_ = os.Setenv("ULTRADNS_USERNAME", "")
+	_ = os.Setenv("ULTRADNS_PASSWORD", "")
+	_ = os.Setenv("ULTRADNS_BASEURL", "")
+	_ = os.Setenv("ULTRADNS_ACCOUNTNAME", "")
+	_, err := NewUltraDNSProvider(endpoint.NewDomainFilter([]string{"test-ultradns-provider.com"}), true)
+	if err != nil {
+		t.Errorf("failed : %s", err)
+	}
+
+	_ = os.Unsetenv("ULTRADNS_PASSWORD")
+	_ = os.Unsetenv("ULTRADNS_USERNAME")
+	_ = os.Unsetenv("ULTRADNS_BASEURL")
+	_ = os.Unsetenv("ULTRADNS_ACCOUNTNAME")
+	_, err = NewUltraDNSProvider(endpoint.NewDomainFilter([]string{"test-ultradns-provider.com"}), true)
+	if err == nil {
+		t.Errorf("expected to fail")
+	}
+}
+
+//zones function test scenario
+func TestUltraDNSProvider_Zones(t *testing.T) {
+	mocked := mockUltraDNSZone{}
+	provider := &UltraDNSProvider{
+		client: udnssdk.Client{
+			Zone: &mocked,
+		},
+	}
+
+	zoneKey := &udnssdk.ZoneKey{
+		Zone:        "",
+		AccountName: "teamrest",
+	}
+
+	expected, _, _, err := provider.client.Zone.SelectWithOffsetWithLimit(zoneKey, 0, 1000)
+	if err != nil {
+		t.Fatal(err)
+	}
+	zones, err := provider.Zones(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(expected, zones) {
+		t.Fatal(err)
```

Comment:
Please use assert.fail function instead of t.fatal everywhere. Please do the similar changes in udsnsdk project.